core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -376,12 +381,36 @@ class Partition(val topic: String,
   }
 
   /**
+   * Check and maybe increment the low watermark of the partition;
+   * this function can be triggered when any replica's logStartOffset changed
```

Comment:
@junrao I see. I just updated the patch to store `lowWatermark` in `Partition` instead of `Replica` and always call `maybeIncrementLeaderLW()` when we retrieve `lowWatermark`. I still mtaintain lowWatermark as state instead of computing it on the fly so that we only needs to call `tryCompleteDelayedRequests` if lowWatermark has increased.